libraries/libfc/include/fc/io/datastream.hpp

```diff
@@ -176,6 +176,38 @@ class datastream<Container, typename std::enable_if_t<(std::is_same_v<std::vecto
    const Container& storage() const { return _container; }
 };
 
+/**
+ * Datastream wrapper that creates a copy of the datastream data read. After reading
+ * the data is available via extract_mirror()
+ * @tparam DataStream datastream to wrap
+ */
+template <typename DataStream>
+class datastream_mirror {
+public:
+   explicit datastream_mirror( DataStream& ds, size_t reserve = 0 ) : ds(ds) {
+      mirror.reserve(reserve);
+   }
+
+   void skip( size_t s ) { ds.skip(s); }
+   bool read( char* d, size_t s ) {
+      if (ds.read(d, s)) {
+         auto size = mirror.size();
+         mirror.resize(size + s);
```

Comment:
This is potentially very inefficient. I suggest `mirror.resize(std::bit_ceil(size + s));`